### PR TITLE
[FLOC-1539] Create container with ports

### PR DESCRIPTION
**Pull request by dwgebler**
Fixes FLOC-1539

### CODE REVIEW

File: docs/advanced/api_examples.yml [hunk 1]
```diff
@@ -270,14 +270,16 @@
```
Comment (itamarst) [inline docs/advanced/api_examples.yml]: I suggest adding an example per feature, given that the ones you're adding now are optional parameters.

File: flocker/control/httpapi.py [hunk 2]
```diff
@@ -437,7 +440,7 @@ def state_datasets(self):
             '$ref': '/v1/endpoints.json#/definitions/configuration_container'},
         schema_store=SCHEMAS
     )
-    def create_container_configuration(self, host, name, image):
+    def create_container_configuration(self, host, name, image, ports=None):
         """
```
Comment (itamarst) [inline flocker/control/httpapi.py]: Update docstring to cover ports.

File: flocker/control/httpapi.py [hunk 3]
```diff
@@ -437,7 +440,7 @@ def state_datasets(self):
             '$ref': '/v1/endpoints.json#/definitions/configuration_container'},
         schema_store=SCHEMAS
     )
-    def create_container_configuration(self, host, name, image):
+    def create_container_configuration(self, host, name, image, ports=None):
```
Comment (itamarst) [inline flocker/control/httpapi.py]: If you have `ports=()` instead you can remove the `if ports:` line.

File: flocker/control/httpapi.py [hunk 4]
```diff
@@ -466,10 +469,31 @@ def create_container_configuration(self, host, name, image):
...
 
+        # If we have ports specified, add these to the Application instance.
+        if ports:
+            application_ports = []
+            for port in ports:
+                application_ports.append(Port(
+                    internal_port=port['internal'],
+                    external_port=port['external']
+                ))
+            application.ports = frozenset(application_ports)
```
Comment (itamarst) [inline flocker/control/httpapi.py]: I am going to make `Application` uses pyrsistent, so it's going to be immutable. I suggest building up a dictionary of additional keyword arguments to pass to `Application()`.

File: flocker/control/httpapi.py [hunk 5]
```diff
@@ -528,6 +552,29 @@ def datasets_from_deployment(deployment):
...
+def container_configuration_response(application, node):
+    """
+    Return a container dict  which confirms to
+    ``/v1/endpoints.json#/definitions/configuration_container``
+
+    :param Application application: An ``Application`` instance.
+    :param unicode node: The host on which this application is running.
+    :return: A ``dict`` containing the container configuration.
+    """
+    image = u'{0}:{1}'.format(
```
Comment (itamarst) [inline flocker/control/httpapi.py]: I assume there's a `DockerImage.to_string`. If there isn't this functionality should be added there.

File: flocker/control/schema/types.yml [hunk 6]
```diff
@@ -45,6 +45,19 @@ definitions:
...
+    title: "Container ports"
+    description: "Mapping of internal to external ports."
+    type: array
+    items:
+      type: object
+      properties:
+        internal:
+          type: integer
+        external:
+          type: integer
```
Comment (itamarst) [inline flocker/control/schema/types.yml]: Maybe also `additionalItems: false` and required list too.

File: flocker/control/test/test_httpapi.py [hunk 7]
```diff
@@ -242,6 +242,115 @@ def test_container_name_collision_different_node(self):
...
+                u'image': u'postgres',
+                u'ports': [{u'internal': 5432, u'external': 54320}]
+            },
+            CONFLICT, {
+                u'description':
+                    u"A specified external port is already in use.",
+            }
+        ))
+        return d
+
```
Comment (itamarst) [inline flocker/control/test/test_httpapi.py]: You should also have test for conflicting with external port on different node.

**Comment by itamarst** [comment]
Thanks! Since this feature is externally testable, you should add an acceptance test ("I add a container with a port, eventually I can connect to that port."). Please address that and the more minor points above and resubmit.

File: flocker/control/test/test_httpapi.py (hunk 7 above)
Comment (dwgebler) [inline flocker/control/test/test_httpapi.py]: Can we not expose the same port number on two different hosts? My first pass was deliberately only checking for an external port conflict on the node you're creating on.

File: flocker/acceptance/test_api.py [hunk 8]
```diff
@@ -231,6 +231,27 @@ def delete_dataset(self, dataset_id):
...
+        :returns: A tuple of (cluster, api_response)
+        """
+        request = post(
+            self.base_url + b"/configuration/containers",
+            data=dumps(properties),
+            headers={b"content-type": b"application/json"},
+            persistent=False
+        )
+
+        request.addCallback(content)
```
Comment (itamarst) [inline flocker/acceptance/test_api.py]: You can use `json_content` to replace `content`+`loads`.

File: flocker/control/httpapi.py [hunk 9]
```diff
@@ -437,7 +441,7 @@ def state_datasets(self):
             '$ref': '/v1/endpoints.json#/definitions/configuration_container'},
```
Comment (itamarst) [inline flocker/control/httpapi.py]: Add new example to list of examples.

File: flocker/control/test/test_schemas.py [hunk 10]
```diff
@@ -53,6 +53,51 @@
...
+            'ports': [{'internal': 80, 'external': '1'}]
+        },
+        # Ports given but invalid key present
+        {
+            'host': '192.168.0.3',
+            'image': 'postgres',
+            'name': 'postgres',
+            'ports': [{'container': 80, 'external': '1'}]
+        },
+        # Ports given but external is not valid intgeger
```
Comment (itamarst) [inline flocker/control/test/test_schemas.py]: "intgeger"

File: flocker/acceptance/test_api.py [hunk 11]
```diff
@@ -312,6 +334,44 @@ def wait_for_cluster(test_case, node_count):
...
+            expected_application = Application(
+                name=data[u"name"],
+                image=DockerImage.from_string(data[u"image"]),
+                ports=frozenset([Port(internal_port=80, external_port=8080)])
+            )
+            node_state = get_node_state(response[u"host"])
+            self.assertIn(data[u"name"], node_state)
+            self.assertEqual(node_state[data[u"name"]], expected_application)
+            self.assertEqual(response, data)
+
```
Comment (itamarst) [inline flocker/acceptance/test_api.py]: This should additionally `loop_until` it can connect to the exposed port.

**Comment by itamarst** [comment]
OK, two minor comments and finish up the acceptance test so it actually connects to the port on the node, then merge when green. Thanks!

**Pull request merged**